---
title: "Introduction to `paramedic`"
author: "Brian D. Williamson and Amy D. Willis"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_vignette:
    keep_md: true
vignette: >
  %\VignetteIndexEntry{Introduction to `paramedic`}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, echo = FALSE, include = FALSE}
library(knitr)
opts_knit$set(cache = FALSE, verbose = TRUE, global.par = TRUE)
```

```{r echo = FALSE}
par(mar = c(5, 12, 4, 2) + 0.1)
```

## Introduction

`paramedic` is a `R` package for estimating microbial concentration. `paramedic` uses information from 16S count data (compositional data on all taxa) and absolute data on a subset of taxa (e.g., qPCR or flow cytometry) to estimate the absolute abundance of all taxa. The method accounts for differing taxon detection efficiencies between the two methods, and produces prediction and confidence intervals as well as point estimates of the absolute abundance of all taxa.

The author and maintainer of the `paramedic` package is [Brian Williamson](https://bdwilliamson.github.io/). For details on the method, check out [our preprint](https://www.biorxiv.org/content/10.1101/761486v1).

## Installation

Currently, the package may only be downloaded and installed from GitHub using the `devtools` package. Type the following commands in your R console to install `devtools` and `paramedic` as needed:

```{r, eval = FALSE}
# install.packages("devtools")
# devtools::install_github("statdivlab/paramedic@main")
```

## Quick Start

This section should serve as a quick guide to using the `paramedic` package -- we will cover the main functions using a data example.

First, load the `paramedic` and `rstan` packages:
```{r}
library("rstan")
library("paramedic")
```

Next, we load in some example data:
```{r}
data(example_16S_data)
data(example_qPCR_data)
```

`example_16S_data` is a matrix of `q = 433` taxa observed from 16S sequencing in `n = 20` samples. We do not have primers to observe the concentration of all taxa, but we can do this for `q_obs = 7` taxa. The observed concentration of the 7 taxa in the 20 samples (in 16S gene copies per unit volume) is in `example_qPCR_data`. These data mimic data observed by the University of Washington (UW) Sexually Transmitted Infections Cooperative Research Center (STICRC). 

We are now ready to estimate concentrations. We pass the data to `run_paramedic`, the workhorse function of the package:

```{r run-paramedic}
## -------------------------------------------------------------
## Obtain estimates of concentration for the first seven taxa
## -------------------------------------------------------------
## this is a small number of iterations, only for illustration
## also, shows how to use control parameters for rstan::stan
mod <- paramedic::run_paramedic(W = example_16S_data[, 1:8], V = example_qPCR_data,
    n_iter = 30, n_burnin = 25, n_chains = 1, stan_seed = 4747,
    control = list(max_treedepth = 15))
mod_summ <- summary(mod$stan_fit, probs = c(0.025, 0.975))$summary
mod_samps <- extract(mod$stan_fit)
```

The output from `run_paramedic` may be manipulated as any output from `rstan::stan` might; however, there are certain posterior summaries that are of particular interest in this problem, namely the posterior mean estimates of concentrations and efficiencies, along with interval estimates for these quantities. The next function extracts these point and interval estimators:
```{r extract-estimates}
## -------------------------------------------------------------
## Extract posterior estimates for the first three taxa
## -------------------------------------------------------------
summs <- paramedic::extract_posterior_summaries(stan_mod = mod_summ,
    stan_samps = mod_samps,
    taxa_of_interest = 1:3, mult_num = 1, level = 0.95,
    interval_type = "wald")
```
and we may finally print them out:
```{r print-summaries}
## -------------------------------------------------------------
## Print out the posterior mean efficiencies and concentrations
## -------------------------------------------------------------
summs$estimates
summs$est_efficiency
```

This output shows that we have point estimates for the true concentration of each taxon within each participant, and the efficiencies of each taxon comparing br16S to qPCR.

## Specifying different hierarchical models

For details on the hierarchical modelling options available in `paramedic`, see [the vignette on hierarchical model specificaion](hierarchical_model_specification.html).

## Running analyses in parallel

`RStan` supports parallelization, which we can use within `paramedic` to speed up computation. In the following example, we re-analyze the data referenced above, but run the models in parallel. Depending on the number of cores on your machine, this can result in significant reductions in computation time.

```{r run-paramedic-parallel}
# parallel::detectCores() will run with multiple cores if available
# follow the steps in the RStan documentation
# (https://github.com/stan-dev/rstan/wiki/RStan-Getting-Started#how-to-use-rstan)
options(mc.cores = parallel::detectCores())
rstan_options(auto_write = TRUE)
## -------------------------------------------------------------
## Obtain estimates of concentration for the first seven taxa
## -------------------------------------------------------------
## this is a small number of iterations, only for illustration
## also, shows how to use control parameters for rstan::stan
mod <- paramedic::run_paramedic(W = example_16S_data[, 1:8], V = example_qPCR_data,
    n_iter = 30, n_burnin = 25, n_chains = 1, stan_seed = 4747,
    control = list(max_treedepth = 15))
mod_summ <- summary(mod$stan_fit, probs = c(0.025, 0.975))$summary
mod_samps <- extract(mod$stan_fit)
```
