<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Run a Stan algorithm to estimate concentrations from combined absolute and relative abundance data — run_paramedic • paramedic</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Run a Stan algorithm to estimate concentrations from combined absolute and relative abundance data — run_paramedic"><meta property="og:description" content="Estimate concentrations (and efficiencies, if applicable) by combining absolute and relative abundance data."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">paramedic</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.3.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/hierarchical_model_specification.html">Hierarchical model specification in `paramedic`</a>
    </li>
    <li>
      <a href="../articles/introduction_to_paramedic.html">Introduction to `paramedic`</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/statdivlab/paramedic/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Run a Stan algorithm to estimate concentrations from combined absolute and relative abundance data</h1>
    <small class="dont-index">Source: <a href="https://github.com/statdivlab/paramedic/blob/HEAD/R/run_paramedic.R" class="external-link"><code>R/run_paramedic.R</code></a></small>
    <div class="hidden name"><code>run_paramedic.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Estimate concentrations (and efficiencies, if applicable) by combining absolute and relative abundance data.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">run_paramedic</span><span class="op">(</span></span>
<span>  <span class="va">W</span>,</span>
<span>  <span class="va">V</span>,</span>
<span>  X <span class="op">=</span> <span class="va">V</span><span class="op">[</span>, <span class="fl">1</span>, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  n_iter <span class="op">=</span> <span class="fl">10500</span>,</span>
<span>  n_burnin <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  n_chains <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  stan_seed <span class="op">=</span> <span class="fl">4747</span>,</span>
<span>  centered <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  inits_lst <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sigma_beta <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">50</span><span class="op">)</span>,</span>
<span>  sigma_Sigma <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">50</span><span class="op">)</span>,</span>
<span>  alpha_sigma <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  kappa_sigma <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  alpha_phi <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  beta_phi <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  sigma_xi <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>W</dt>
<dd><p>The relative abundance data, e.g., from broad range 16S sequencing with "universal" primers. Expects data (e.g., matrix, data.frame, tibble) with sample identifiers in the first column. Sample identifiers must be the same between W and V, and the column must have the same name in W and V.</p></dd>


<dt>V</dt>
<dd><p>The absolute abundance data, e.g., from taxon-specific absolute primers. Expects data (e.g., matrix, data.frame, tibble) with sample identifiers in the first column. Sample identifiers must be the same between W and V, and the column must have the same name in W and V.</p></dd>


<dt>X</dt>
<dd><p>The covariate data. Expects data (e.g., matrix, data.frame, tibble) with sample identifiers in the first column. Sample identifiers must be the same between W, V, and X, and the column must have the same name in W, V, and X. If X only consists of the subject identifiers, then no covariates are used.</p></dd>


<dt>k</dt>
<dd><p>the number of batches that the relative abundance data W were analyzed in. If k = 0 (the default), then batch effects are not considered.</p></dd>


<dt>n_iter</dt>
<dd><p>The total number of iterations per chain to be run by the Stan algorithm. Defaults to 10500.</p></dd>


<dt>n_burnin</dt>
<dd><p>The total number of warmups per chain to be run by the Stan algorithm. Defaults to 10000.</p></dd>


<dt>n_chains</dt>
<dd><p>The total number of chains to be run by the Stan algorithm. Defaults to 4.</p></dd>


<dt>stan_seed</dt>
<dd><p>The random number seed to initialize.</p></dd>


<dt>centered</dt>
<dd><p>If <code>TRUE</code>, uses a centered parameterization on the true concentrations \(\mu\); if <code>FALSE</code> (the default), uses a non-centered parameterization.</p></dd>


<dt>inits_lst</dt>
<dd><p>An optional list of initial values of the parameters. Must be a named list; see <code><a href="https://mc-stan.org/rstan/reference/stan.html" class="external-link">stan</a></code>.</p></dd>


<dt>sigma_beta</dt>
<dd><p>Hyperparameter specifying the prior variance on \(\beta_0\). Defaults to \(\sqrt{50}\).</p></dd>


<dt>sigma_Sigma</dt>
<dd><p>Hyperparameter specifying the prior variance on \(\Sigma\). Defaults to \(\sqrt{50}\).</p></dd>


<dt>alpha_sigma</dt>
<dd><p>Hyperparameter specifying the shape parameter of the prior distribution on \(\sigma_e\). Defaults to 2.</p></dd>


<dt>kappa_sigma</dt>
<dd><p>Hyperparameter specifying the scale parameter of the prior distribution on \(\sigma_e\). Defaults to 1.</p></dd>


<dt>alpha_phi</dt>
<dd><p>Hyperparameter specifying the shape parameter of the prior distribution on \(\phi\). Defaults to 0; a negative binomial model can be specified if both <code>alpha_phi</code> and <code>beta_phi</code> are nonzero.</p></dd>


<dt>beta_phi</dt>
<dd><p>Hyperparameter specifying the rate parameter of the prior distribution on \(\phi\). Defaults to 0; a negative binomial model can be specified if both <code>alpha_phi</code> and <code>beta_phi</code> are nonzero.</p></dd>


<dt>sigma_xi</dt>
<dd><p>Hyperparameters specifying the variance of efficiencies over batches. Only used if <code>k</code> is greater than zero. Defaults to 1.</p></dd>


<dt>...</dt>
<dd><p>other arguments to pass to <code><a href="https://mc-stan.org/rstan/reference/stanmodel-method-sampling.html" class="external-link">sampling</a></code> (e.g., control).</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>An object of class <code>stanfit</code>.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>We fit a hierarchical model in Stan to the data, with goal to estimate true concentration for all taxa. There are two available hierarchical models. The two only differ in the way that the true concentration is parameterized. While these hierarchical models are mathematically identical, they are fit differently by Stan. The first hierarchical model is a "centered" model, where $$\beta_0 ~ N(0, \sqrt{50})$$ $$\log \Sigma ~ N(0, \sqrt{50})$$ $$\log \mu ~ N(\beta_0 + X\beta_1, \Sigma).$$ We call this model <code>variable_efficiency_centered.stan</code>. The second hierarchical model is a "noncentered" model, where $$\beta_0 ~ N(0, \sqrt{50})$$ $$\log \Sigma ~ N(0, \sqrt{50})$$ $$\log \gamma ~ N(0, I)$$ $$\log \mu = \sqrt{\Sigma} \log \gamma + \beta_0 + X\beta_1.$$ We call this model <code>variable_efficiency.stan</code>.
In most cases, we suggest using the noncentered model. However, if the model does not converge in a reasonable amount of time using the noncentered model, consider trying the centered model.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="https://mc-stan.org/rstan/reference/stan.html" class="external-link">stan</a></code> and <code><a href="https://mc-stan.org/rstan/reference/stanmodel-method-sampling.html" class="external-link">sampling</a></code> for specific usage of the <code>stan</code> and <code>sampling</code> functions.</p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># load the package, read in example data</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://statdivlab.github.io/paramedic">"paramedic"</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">example_16S_data</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">example_qPCR_data</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># run paramedic (with an extremely small number of iterations, for illustration only)</span></span></span>
<span class="r-in"><span><span class="co"># on only the first 10 taxa</span></span></span>
<span class="r-in"><span><span class="va">mod</span> <span class="op">&lt;-</span> <span class="fu">run_paramedic</span><span class="op">(</span>W <span class="op">=</span> <span class="va">example_16S_data</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>, V <span class="op">=</span> <span class="va">example_qPCR_data</span>,</span></span>
<span class="r-in"><span>n_iter <span class="op">=</span> <span class="fl">30</span>, n_burnin <span class="op">=</span> <span class="fl">25</span>, n_chains <span class="op">=</span> <span class="fl">1</span>, stan_seed <span class="op">=</span> <span class="fl">4747</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SAMPLING FOR MODEL 'variable_efficiency' NOW (CHAIN 1).</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: Gradient evaluation took 0.000125 seconds</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: 1000 transitions using 10 leapfrog steps per transition would take 1.25 seconds.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: Adjust your expectations accordingly!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: WARNING: There aren't enough warmup iterations to fit the</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1:          three stages of adaptation as currently configured.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1:          Reducing each adaptation stage to 15%/75%/10% of</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1:          the given number of warmup iterations:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1:            init_buffer = 3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1:            adapt_window = 20</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1:            term_buffer = 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: Iteration:  1 / 30 [  3%]  (Warmup)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: Iteration:  3 / 30 [ 10%]  (Warmup)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: Iteration:  6 / 30 [ 20%]  (Warmup)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: Iteration:  9 / 30 [ 30%]  (Warmup)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: Iteration: 12 / 30 [ 40%]  (Warmup)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: Iteration: 15 / 30 [ 50%]  (Warmup)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: Iteration: 18 / 30 [ 60%]  (Warmup)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: Iteration: 21 / 30 [ 70%]  (Warmup)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: Iteration: 24 / 30 [ 80%]  (Warmup)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: Iteration: 26 / 30 [ 86%]  (Sampling)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: Iteration: 28 / 30 [ 93%]  (Sampling)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: Iteration: 30 / 30 [100%]  (Sampling)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1:  Elapsed Time: 0.016 seconds (Warm-up)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1:                0.002 seconds (Sampling)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1:                0.018 seconds (Total)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Chain 1: </span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>There were 1 chains where the estimated Bayesian Fraction of Missing Information was low. See</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> https://mc-stan.org/misc/warnings.html#bfmi-low</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Examine the pairs() plot to diagnose sampling problems</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The largest R-hat is 2.5, indicating chains have not mixed.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> Running the chains for more iterations may help. See</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> https://mc-stan.org/misc/warnings.html#r-hat</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Brian D. Williamson, James Hughes, Amy Willis.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

